esp/platform/espcfg.cpp

```diff
@@ -236,44 +248,58 @@ void CEspConfig::ensureSDSSession()
         processSessionTree = sessionRoot->addPropTree(PathSessionProcess);
         processSessionTree->setProp("@name", m_process);
     }
+    return processSessionTree;
+}
 
-    ensureSDSSessionApplications(processSessionTree);
-    sdsSessionEnsured = true;
+void CEspConfig::initSDSSessionCleaner(bool detachedFromDali)
+{
+    if (m_sessionCleaner)
+        return;
 
     if (serverSessionTimeoutSeconds != ESP_SESSION_NEVER_TIMEOUT)
     {
         VStringBuffer espSessionSDSPath("%s/%s[@name=\"%s\"]", PathSessionRoot, PathSessionProcess, m_process.str());
         Owned<IPropertyTree> proc_cfg = getProcessConfig(m_envpt, m_process);
-        m_sessionCleaner.setown(new CSessionCleaner(espSessionSDSPath.str(), proc_cfg->getPropInt("@checkSessionTimeoutSeconds",
-            ESP_CHECK_SESSION_TIMEOUT)));
+        m_sessionCleaner.setown(new CSessionCleaner(espSessionSDSPath.str(), detachedFromDali,
+            proc_cfg->getPropInt("@checkSessionTimeoutSeconds", ESP_CHECK_SESSION_TIMEOUT)));
         m_sessionCleaner->start();
     }
 }
 
-void CEspConfig::ensureSDSSessionApplications(IPropertyTree* espSession)
+void CEspConfig::ensureSDSSessionApplications(IPropertyTree* espSession, unsigned short port)
 {
-    std::list<int> bindingPorts;
-
-    for (const auto pbfg: m_bindings)
+    VStringBuffer appStr("%s[@port=\"%d\"]", PathSessionApplication, port);
+    IPropertyTree* appSessionTree = espSession->queryBranch(appStr.str());
+    if (!appSessionTree)
     {
-        auto it = std::find(bindingPorts.begin(), bindingPorts.end(), pbfg->port);
-        if (it == bindingPorts.end())
-            bindingPorts.push_back(pbfg->port);
+        IPropertyTree* newAppSessionTree = espSession->addPropTree(PathSessionApplication);
+        newAppSessionTree->setPropInt("@port", port);
     }
+}
+
+void CEspConfig::addBindingForSDSSession(unsigned short port)
+{
+    if (!m_sessionCleaner) //No session needed or no dali
+        return;
+
+    auto it = std::find(bindingPorts.begin(), bindingPorts.end(), port);
```

Comment:
If this can be called on multiple threads (I think it can from Tony's comment, and the EsdlNotifyType::BindingAdd processing.  Then bindingPorts also needs to be protected by the critical section.  (Otherwise this code could crash.)

Reply:
Will fix